Jetty 10.0.x quic http3

gregw opened the pull request:
This PR is a draft that I created against the h3 development branch so we can use github tools to review/discuss the code

lachlan-roberts committed:
Copy http2-hpack files into http3-qpack.

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
Refactor names of HPACK to QPACK

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
Extract the StaticTable from the QpackContext.

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
Decouple the DynamicTable from the QpackContext.

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
Add copy of the new QPACK static table from SPEC.

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
Implement a decoder instruction parser & add tests.

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
Implement setCapacity and duplicate on the Encoder Instruction Parser.

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
Implement insertNameWithReference and insertWithLiteralName in EncoderInstructionParser.

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
Implement Encoder & Decoder Streams to encode instructions.

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
Use classes for encoder and decoder instructions.

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
Use setPrefix API for NBitIntegerParser

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
Add basic implementation for the QpackDecoder.

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
QpackDecoder should take the Stream ID to decode.

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
Remove the QPACK tests which only apply to HTTP/2 HPACK.

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
Write the basic implementation for the QpackEncoder.

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
Work through examples B.1. and B.2. from spec and fix bugs.

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
Rewrite QpackEncoder.encode for HttpField, implement Duplicates.

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
Implement SectionAcknowledgment and StreamCancellation

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
wip

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
Add synchronization to both Qpack Encoder and Decoder.

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
Allow extensions of QpackEncoder to override choice to huffman encode.

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
DynamicTable should use a configurable draining index, only updated on insertions.

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
Implement the QpackFieldPreEncoder allowing it to use static table references.

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
The QpackEncoder should be able to use PreEncodedHttpFields.

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
Renaming and change package access.

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
Qpack refactoring, testing and bug fixes.

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
Use calculated header size for QpackEncoder buffer allocation.

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
Refactor QPACK to use an internal package for non API classes.

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
Fix broken test case EncodeDecodeTest.

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lorban committed:
fix default value of _currentInterestOps

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
dispatch session readable events to EWYK

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
fix a couple connection closing bugs

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
fix buffers clearing and releasing on looping

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
force loading libquiche early

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
add protection against using a freed quiche connection

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
improvements about how tasks are dispatched

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
change test logs to INFO

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
add comment

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
minor refactoring

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
move the execution strategy from connection to session

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
code alignment

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
improve pkcs12 -> pem convertor

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
replace InetSocketAddress decoding/encoding with side-stack argument passing

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
replace InetSocketAddress decoding/encoding with side-stack argument passing

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
fix lifecycle on unrecoverable network error

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
introduce IClientConnector interface

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
start quic client plumbing

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
add big file download test

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
better align client and server impls

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
kinda sorta done the piping work up to Session.process()

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
fix NPE during dispose

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
create stream 0 when client connection is set up

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
HttpDestination can now be told to use cleartext even when the scheme is secure

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
cleanup SelectorManager connect FSM

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
rename client datagram connector

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
cleanup SelectorManager accept FSM

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
move datagram endpoint to common module

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
cleanup

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
better align client and server quic connections

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
better align client and server quic sessions

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
make client use common classes

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
make server use common classes

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
remove unneeded parameter

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
force selector count to 1

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
rename connectors

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
make quic token minting and validation pluggable

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
fix checkstyle

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
do not close when QuicConnection gets an exception as it is shared by multiple quic connections

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
make onClose overridable and null-safe

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
add a way to build a connection id object from a cid array

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
notify and cleanup when quic connection is refused

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
throw a specific exception on token validation error

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
introduce char_pointer native type

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
clean up alpn protocol configuration

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
clean up alpn protocol configuration

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
more end2end tests

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
fix max multiplex

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
client quic config is kinda configurable now

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
make ClientQuicConnector extend from ClientConnector

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
improve quiche error handling

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
get rid of HttpClientTransportOverQuic and IClientConnector

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
get rid of QuicheConfig in QuicClientConnector

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
cleanup tests

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
cleanup tests

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
todos and cleanups

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
todos and cleanups

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
todo

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
fix log

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
fix log

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
do not write directly to the endpoint

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
improve toString

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
do not write directly to the endpoint

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
dispatch from the timer thread to iterate

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
improve debug logs

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
make close(Throwable) private

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
improve debug logs

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
minor ref change

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
refactor session creation

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
get rid of promoteSession()

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lachlan-roberts committed:
Use MetaData instead of HttpFields for QPACK

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
Filter out connection specific headers for Http3Fields

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
QpackEncoder.encode should take buffer to encode into instead of allocating.

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
Remove all synchronization from QpackEncoder & QpackDecoder

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
Improve QPACK testing for instruction parsers.

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
Fix licence header, and checkstyle issues.

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
Improvements & bug fixes from testing

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
Cleanups, bug fix and improve testing

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

sbordet committed:
Issue #6728 - QUIC and HTTP/3

- Started implementation of HTTP/3 APIs, frames and parsing.
- Started implementation of server-side HTTP/3.
- Reworked the threading model in QUIC to be more efficient and avoid contending on the QPACK encoder/decoder.
- Updated QPACK NBitLongEncoder to support QUIC stream IDs of type long.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

gregw commented on jetty-quic/quic-common/src/main/java/org/eclipse/jetty/quic/common/ProtocolQuicSession.java:
```diff
@@ -0,0 +1,56 @@
+
+import java.util.concurrent.atomic.AtomicLong;
+
+public class ProtocolQuicSession
```
@sbordet @lorban We should do a 'Http1ProtocolQuicSession` so we can quickly validate the basic lifecycle with working code.

gregw commented:
Hmmmm this PR might be a bit big to be useful... but we will ultimately need to review, so starting early???

However, I'm wondering if we can reduce verbosity a little bit by squashing the many commits (yeah I know I normally argue against squashes in branches, but rules always have exceptions)?

gregw commented on jetty-http3/http3-server/src/main/java/org/eclipse/jetty/http3/server/HTTP3QuicSession.java:
```diff
@@ -0,0 +1,49 @@
+
+public class HTTP3QuicSession extends ProtocolQuicSession
+{
+    private final QpackDecoder decoder;
+    private final Instruction.Handler decoderHandler = new QpackDecoderInstructionHandler();
```
So it is expected that there will be an encoder here as well.
Potentially also parsers/generators, Flushers and queues of frames????

gregw commented on jetty-http3/http3-qpack/src/main/java/org/eclipse/jetty/http3/qpack/QpackDecoder.java:
```diff
@@ -0,0 +1,276 @@
+    private final List<EncodedFieldSection> _encodedFieldSections = new ArrayList<>();
+    private final NBitIntegerParser _integerDecoder = new NBitIntegerParser();
+    private final int _maxHeaderSize;
+    private MetaData _metaData;
```
@lachlan-roberts 
A `Decoder` can decode `MetaData` for multiple streams, so why is there a single `_metaData` field?  I'm guessing this is the current meta data field being built?  But are we sure that we can't be called like:
```java
    decoder.decode(streamA, bufferWithFirstHalfOfMetaDataA, myHandler);
    decoder.decode(streamB, bufferWithFirstHalfOfMetaDataB, myHandler);
    decoder.decode(streamA, bufferWithSecondHalfOfMetaDataA, myHandler);
    decoder.decode(streamB, bufferWithSecondHalfOfMetaDataB, myHandler);
```
If this can't happen, then at the very least this needs to be documented, but I'd also rather we remembered the streamId in progress and threw an ISE if we are called with another stream when we have not finished the first.

gregw commented on jetty-http3/http3-qpack/src/main/java/org/eclipse/jetty/http3/qpack/QpackDecoder.java:
```diff
@@ -0,0 +1,276 @@
+        void onMetaData(long streamId, MetaData metadata);
+    }
+
+    public boolean decode(long streamId, ByteBuffer buffer, Handler handler) throws QpackException
```
@lachlan-roberts I've given this feedback before.... but I'll try again as I think I'm getting my objections a bit better formed.

I don't like the API of `QpackDecoder` as it looks like an active object rather than a passive state machine (like SslEngine or quiche).    This is important because the caller of `decode` is likely to be in a very specific invocation context, specifically needing to know if it is BLOCKING, NON_BLOCKING or EITHER, but with this API it is not possible to say that `decode` is a non blocking call or not, since it depends entirely on the implementations of the handlers that the decoder has.   Now it may be possible for the user of the decoder to know completely the invocation behaviours of those call backs, but even so that relationship may not be clear to somebody reading the code.

A simple example is that if the `Handler.onMetaData` callback goes on the actually invoke the application with the request that is represented by that metadata, then a thread may be still inside the decoder and blocked (or active) in the application when another thread needs to call `decode` either for other streams or perhaps for trailers on the same stream.   So sure, we can say that the handler should not be implemented that way... but it would be better to have an API that couldn't be implemently badly.

To take control of the invocation, the user of `Decoder` will need to:
 + construct the `Decoder` with an `Instruction.Handler` that never blocks, so it probably needs to just add them to a queue.
 + call the `decode` method with a `Handler` that never blocks and just remembers the `MetaData`
 + reacts to a true return from `decode` by looking for a remembered `MetaData`  and acting on it.
 + ensures all remembered `Instructions` are handled after the `decode` is returned but not delayed by any metadata handling.

Why not just make that explict with `decode` return a `Result` that contains both  `MetaData` and a collection of `Instructions`.  Or alternately just an action enumeration saying that there is a `MetaData` and/or collection of `Instructions` available, which can then be pulled from the decoder.

gregw commented on jetty-http3/http3-qpack/src/main/java/org/eclipse/jetty/http3/qpack/QpackDecoder.java:
```diff
@@ -0,0 +1,276 @@
+        return metaData != null;
+    }
+
+    void parseInstruction(ByteBuffer buffer) throws QpackException
```
@lachlan-roberts is this public API?  is this the only way that recieved instructions make it into the instruction parser?

gregw commented on jetty-http3/http3-qpack/src/main/java/org/eclipse/jetty/http3/qpack/QpackDecoder.java:
```diff
@@ -0,0 +1,276 @@
+    /**
+     * This delivers notifications from the DecoderInstruction parser directly into the Decoder.
+     */
+    class DecoderAdapter implements DecoderInstructionParser.Handler
```
@lachlan-roberts  
I don't think `Adaptor` is the right name here.   That implies it is an abstract implementation that is intended to be further extended.
Rather this is the implementation of the parser handler that integrates into the `Decoder`, so perhaps just call it `InstructionHandler`

gregw committed:
Added puml diagram

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw commented on jetty-quic/quic-server/src/main/java/org/eclipse/jetty/quic/server/H3Server.puml:
```diff
@@ -0,0 +1,65 @@
+@startuml
+title "H3 Quic Server"
```
@sbordet @lachlan-roberts @lorban I know it is early days, but I do think puml diagrams like this will help us work out details.
I'm happy to mostly maintain it, but if you can review and comment.

lachlan-roberts committed:
Issue #6728 - Fixes for QPACK

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lorban committed:
#6730: fix ip encoding endianness + move timespec to ffi

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

sbordet committed:
Issue #6728 - QUIC and HTTP/3

- Continued implementation of HTTP/3 APIs, frames and parsing, client and server.
- Figured out creation of streams, to support special HTTP/3 streams.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Issue #6728 - QUIC and HTTP/3

- Implemented handling of the control stream and SETTINGS frame/parser/generator.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

gregw commented:
@sbordet @lorban @lachlan-roberts On the call last week I struggled to make my concerns clear about the scheduling of h3.     I've put more thought into it, so let me have another try.

First, let's go over how the execution strategy works in other places:
 + For the selector manager, we don't have a thread do a select and then iterator over all the selected keys.  Instead we have a producer model that when asked to produce, the producer may look at the results from the previous select and produce a task that is associated with one of the selected keys.  The producer doesn't know what protocol is being spoken and a selected key could be an acceptor, h1, h2, ssl, or even just an update task could be produced.
 + At the h2 level, we have another execution strategy where the producer does the reading/parsing of the frames.  It produces a single tasks even though there may be more reading able to be done or more frames in the data already read.   This is key because we want the execution strategy to have the best change for using the same thread that did the reading & parsing to als odo the handling.

So in both cases, we have a producer that unwinds a natural loop (over selected keys, or over fill&parsed frames) and produces a single task that is hot in the CPU cache that can be fed to the execution strategy to potentially be directly executed without dispatch.

For quic+h3 there is an equivalent to the selector selecting a set of keys.   It is quiche producing a set of readable streams and writable streams.   We need a producer at that level that will produce the next tasks from that list - which may be to pick a single readable stream, read the data, pick the correct parser for that (based on streamID it could be settings stream, an instruction stream or a H3 stream(which itself could parse header frames or data frames).  A single task should be returned from that production so that the same thread can go onto handle that task with the frame/parsing hot in its cache.    The next producing will then produce the next task - which may be from the same read/parse (eg there may be multiple metadata's produced from a single instruction), or it may go to the next readable streamId to do a read/parse etc.

Ultimately, a produce will find nothing to do and return, leaving that execution strategy idle, to be woken up when it is next selected.   Note that this should allows the execution strategy itself to do the "1 and only 1 thread" logic that @sbordet has currently written around AtomicInt.  The ExecutionStrategy already only ever allows a single producer, so that will protect our qpack tables and parsers from having multiple threads - we don't need to do anything extra.

OK, so it  is a little more complex that the h2 execution stragegy, in that we don't have a single parser that the producer can feed.  But I'm pretty sure we can do it similar to the selector, where the specifics of how to act  are in the key attachment.  For quiche/h3 it will be the streamId that will allows a sub-producer to be selected by the selector, that will then do what ever it needs to do to produce the next task (read & parse,  or produce task from a previous parse, or wake up a writer, or callback the application to say it can write, etc. etc.)

The key think is that the execution strategy should be thought of as having a producer at the QuicSession level and that it pulls tasks out of the multiple streams that may be indicated as active by quiche.    Currently we are thinking that the execution strategy is after the stream layers and that QuicSession pushes to the streams and those streams push to the producer of a subsequent execution strategy.   I don't think that can work... and if it can, it is too complex.   I'm sure we can "push into a producer", but I think it will involve handoffs.  Much better to have a producer that pulls tasks out of the various statemachines and parsers.

@lachlan-roberts, so to fit this model, that is why I like a Decoder that you pull from rather than it actively calls handlers.     There will be a producer that will ask the Decoder: "do you have any thing I can act on?" and the Decoder can return an Instruction to send or an emitted MetaData.  The producer can act on that task and arrange for another producer to call the Decoder again.  Only once the Decoder says: "Nothing more I can do", does the producer look to see if the instruction stream and/or normal stream is able to be read from, and the resulting data parsed into frames and fed into the Decoder, which is then asked agin: "do you have anything I can act on?".  Doing this way, will unwind the loops that currently exist in the Decoder and allow the execution strategy to be involved.  More over, whilst there is probably some queue of metadata & instances that will be produced, this queue can be internal, final and reused.   A new list will not need to be created on every parse.

sbordet committed:
Issue #6728 - QUIC and HTTP/3

- Restructured code
- Split implementation of control and message parsing/generation.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

gregw commented on jetty-quic/quic-common/src/main/java/org/eclipse/jetty/quic/common/QuicSession.java:
```diff
@@ -0,0 +1,415 @@
+
+        quicheConnection.feedCipherText(cipherBufferIn, remoteAddress);
+
+        if (quicheConnection.isConnectionEstablished())
```
maybe first check for `protocolSession!=null` as we don't know if/when the established state might revert to false during a shutdown?

gregw committed:
Added puml diagram

More execution strategy ideas

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw committed:
Added puml diagram

More execution strategy ideas

Signed-off-by: Greg Wilkins <gregw@webtide.com>

lorban committed:
add macos x86_64 and aarch64 quiche 0.9.0 universal binary lib

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
split linux and macos native structures to account for native differences

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
keep quiche version check out of if (debug enabled) branch

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lachlan-roberts committed:
Improve exceptions from QPACK

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lorban committed:
add IPv6 support

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
generify sockaddr and sockaddr_storage to simplify native mappings

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

sbordet committed:
Issue #6728 - QUIC and HTTP/3

- Fixed control streams, must be unidirectional.
- Initial implementation of events to HTTP/3 APIs.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Issue #6728 - QUIC and HTTP/3

- Simple request/response (no content) working.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

gregw commented on jetty-io/src/main/java/org/eclipse/jetty/io/EndPoint.java:
```diff
@@ -203,7 +203,15 @@ default void close()
+        throw new UnsupportedOperationException();
+    }
+
+    default SocketAddress receive(ByteBuffer buffer) throws IOException
+    {
+        throw new UnsupportedOperationException();
+    }
```
I'm not convinced it is correct to add this method to our abstract EndPoint.  OK for the short term, but perhaps we should consider refactoring so that `EndPoint` has neither `fill` nor `receive`.  Instead these signatures should be moved to `DatagramEndPoint` and `StreamEndPoint`?   Ditto for `flush`,  `write` and `send`.

But let's leave it this way for now and just be aware that it may need to change.

lachlan-roberts committed:
Add javadoc to public methods on QPACK encoder & decoder.

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

gregw commented on jetty-quic/quic-common/src/main/java/org/eclipse/jetty/quic/common/QuicSession.java:
```diff
@@ -0,0 +1,434 @@
+ * <p>On the receive side, a QuicSession <em>fans-out</em> to multiple {@link QuicStreamEndPoint}s.</p>
+ * <p>On the send side, many {@link QuicStreamEndPoint}s <em>fan-in</em> to a QuicSession.</p>
+ */
+public abstract class QuicSession
```
@sbordet @lorban  I find the naming very confusion.   A `QuicSession` containers a `QuicheConnection` and the `QuicConnection` is not a connection at all, but more part of the `Connector` or maybe the `ConnectionFactory`.

I think it would be far simpler to rename `QuicSession` to `QuicConnection`, as that is what it is.  The current `QuicConnection` is only named that way because it is being shoehorned into the current SelectorManager and AbstractConnection.  It kind of fits, but the wrong name shows that there is probably a better way.

lachlan-roberts committed:
Improve javadoc and exception handling for http3-qpack

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
Fix licence header in quic StreamType

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

sbordet committed:
Issue #6728 - QUIC and HTTP/3

- Implemented DATA frame generation and parsing.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

gregw commented:
Some more musings on execution strategy.... this time mostly for my own consumption....

Currently the reading of h3 datagrams is looking like the following pseudo code:
```java
SocketAddress peer = DatagramChannel.receive(netbuf)
QuicheConnectionId cId = QuicheConnectionId.fromPacket(netbuf)
QuicheConnection quicheConnection = map.get(cId);
if (quicheConnection==null)
{
    QuicheConnection quicheConnection = QuicheConnection.tryAccept(config, new SimpleTokenValidator(peer), netBuf, peer);
    if (quicheConnection==null)
    {
        if (QuicheConnection.negotiate(new SimpleTokenMinter(peer), netBuf, negBuf))
            datagramChannel.send(negBuf, peer);
        return null;
    }
}
quicheConnection.feedCipherText(netBuf, peer); // TODO what if not all consumed?

// At this point current code mutually excludes and execute the following:
for (streamId : quicheConnection.readableStreamIds())
{
    quicheConnection.drainClearTextForStream(streamId, streamBuf);
    H3Parser parser = pickParserBy(streamId); // In reality done by picking a ProtoQuicStream instance
    H3Frame frame = parser.parse(streamId, streamBuf)

    if (frame instanceof HeaderFrame)
    {
        decoder.decode(frame.getContent());
        while (qpacDecoder.hasMetaData())
        {
            request.setMetData(qpacDecoder.takeMetaData();
            if (httpChannel.onRequest(request))
                execute(httpChannel);
        }
    }
    else if (frame instanceof DataFrame)
    {
        if (httpChannel.onContent(frame.getContent()))
            execute(httpChannel);
    }
    else if (frame instanceof Instruction)
    {
        decoder.parseInstruction(frame.getContent());
        while (qpacDecoder.hasMetaData())
        {
            request.setMetData(qpacDecoder.takeMetaData();
            if (httpChannel.onRequest(request))
                execute(httpChannel);
        }
    }
    else if (frame instanceof Settings)
    {
    }
}
```
This is probably triple executed: 1) the default onFillable mechanism assumes blocking so the selectorManger strategy is effectively an execute; 2) the execute before iterating over the readable streams; 3) the HttpChannel will need to be executed.

With regards to parallelism:
 + the `QuicConnection.onFillable` will only be called by a single thread, so only a single `feedCipherText` call will happen at once.  Ideally we should be able to do `feedCipherText` for different connectionIds in parallel
 + once fed the cipherText, the iterating over readable streams is executed, so that iteration can be done for multiple connections at once, but each stream within a connection is processed serially until the HttpChannel is executed.  This means that the channel will never be handled by the same thread that did the reading and parsing.

So I think this is suffering from being a square peg shoved into the round hole that is our SelectorManager->EndPoint->Connection abstraction.   It looks like it might work, but I don't think it will ever work efficiently.   So I'm going to think about how this could/should look if we started from scratch and tried to directly apply our execution strategy producer/consumer abstraction.
 
We need to well match our execution strategies with the 2 fan outs: multiple connections; and  multiple streams within a connection:
 + The initial `receive` and any `tryAcquire` calls should be done in the producer of the first level ExecutionStrategy.
+ The `quicheConnection.feedCipherText` should be done withing a consumer of the first level ExecutionStrategy, so they can happen in parallel for different connections.   The trick will be what to do if two tasks for the same connection are produced quickly? Should the second one block and wait for the first, or just queue its data and return?  I'm favouring the later as the consumer will probably just be the second level execution strategy, so the task becomes `{ connection.addDatagram(buf); connection.produce();}`. Actually the add can probably be done in the first level produce so the consume may just be to call the 2nd level produce().
+ There may need to be a queue of encrypted data buffers in each connection.  Maybe there is a queue inside quiche, but the call to `feedCipherText`  may not be fully consuming, so we'd need to store the unconsumed buffer anyway.
+ The 2nd level producer will do `feedCipherText`, frame parsing and header decoding for a connection. Since producers are single threaded, this means only a single thread accessing the decoder and other associated state at once.
+ The 2nd level consumer tasks will be the HttpChannel, so they can run in parallel if there are many, but by the same producing thread if there is only 1 active.

I think it is possible to have the same thread: receive; feedCipherText; parse; decode; HttpChannel.run();   2 levels of producers will allow other threads to take over producing at the receive step and/or the feedCipeherText step.

lorban committed:
#6730 add support for windows x86_64 quiche 0.9.0 binary lib

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
add method to get congestion window capacity

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
rename methods to get congestion window capacity

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

gregw commented:
@sbordet @lachlan-roberts @lorban Oh oh!  I think we have a problem with back pressure.

If I understand quic/quiche correctly (and there is a good chance that I don't) then back pressure on the read side is done by not calling `quicheConnection.drainClearTextForStream`.  Ie. once you have  called drain, then those bytes are considered consumed and flow control magic will happen to allow more bytes to arrive.

Thus our current approach has been wrong - we are reading those bytes on the selector callback, parsing them and then looking to use an invocation strategy to invoke the application.   But that can't work because what if the application has not yet done a read?     I think this will have to be more like h1 and that the call to drain will be done within the application thread calling `read`, which will call `produceContent` which will call `drainClearTextForStream` and then parse those bytes for data (or trailers!!!).     This means that `needContent` will need to register something, so that when the stream is next seen in the `readableStreamIds` then the only action taken is to call `onDataAvailable` and allow the application to again call `read` which will ultimately call `drainClearTextForStream`

lachlan-roberts committed:
Fix JPMS warning from QpackDecoder

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
Refactor QpackEncoder and QpackDecoder for simplicity.

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lorban committed:
fix IPv6

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

sbordet committed:
Issue #6728 - QUIC and HTTP/3

- Changed HTTP/3 parsers to return the Frame rather than notifying a listener.
- Correctly linked encoder and decoder streams to respectively QpackEncoder and QpackDecoder.
- Fixed buffer release.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

lorban committed:
replace synchronized keyword with AutoLock

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

sbordet committed:
Issue #6728 - QUIC and HTTP/3

- Fixed handling of the last frame, exposing a new API in QuicStreamEndPoint.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Issue #6728 - QUIC and HTTP/3

- Changed the parsers to emit during parsing, rather than returning the event.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

lorban committed:
add low-level quiche tests

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

sbordet committed:
Issue #6728 - QUIC and HTTP/3

- Implemented data support.
- Implemented trailer support.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Issue #6728 - QUIC and HTTP/3

- Updates to the data support after review.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

lachlan-roberts committed:
update qpack parseInstructionBuffer name

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
Implement stream cancellation in HTTP/3 QPACK

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

gregw commented:
@sbordet can you help me fill out the following breakdown of all the layers involved and the basics of how they communicate?:

 +  SelectorManager <-> DatagramChannelEndPoint: 
    * -> onSelected()
    * <- submit(selectorUpdate)
 +  DatagramChannelEndPoint <-> QuicConnection:
    * -> onFillable()
    * <- fillInterested(), receive(buffer), send(address, buffers), write(callback, address, buffers)
 + QuicConnection <-> QuicSession:
    * -> process(address, buffer)
    * <- write(callback, address, buffers)
 + QuicSession <-> [Client|Server]HTTP3Session (ProtocolSession)
    * -> process();
    * <- getReadableStreamIds(), getWritableStreamIds()
 +  [Client|Server]HTTP3Session <-> QuicStreamEndPoint:
    * -> onReadable(), onWritable()
    * <- fill(streamId, buffer), flush(streamId, buffer)
 + QuicStreamEndPoint <-> HTTP3StreamConnection:
    * -> onFillable()
    * <- fillInterested(), fill(buffer), flush(buffer), write(callback, buffer)
 +  HTTP3StreamConnection <-> HTTP3Session:
    * -> parser.parse(buffer) -> onXYZ() parser events
    * <- writeFrame(streamId, frame, callback)
 +  HTTP3Session <-> HTTP3Stream: 
    * -> onXYZ(Stream) events (e.g. onRequest(), onResponse(), onTrailer(), onDataAvailable(), etc.)
    * <- writeFrame(streamId, frame, callback)
 +  HTTP3Stream <-> HttpChannelOverHTTP3: 
    * -> onXYZ(Stream) events
    * <- readData(), demand(), writeFrame(streamId, frame, callback)
 + HttpChannelOverHTTP3 <-> Servlet:
    * -> Handler.handle(), onContentProducible(),  sendCallback.succeeded();
    * <- needContent(), produceContent(), transport.send(request, response, buffer, last, callback);
 + Servlet <-> App:
    * -> service(req,resp), onDataAvailable(), onWritePossible()
    * <- in.isReady(), in.read(byte[]), out.isReady(), out.write(byte[])

sbordet committed:
Issue #6728 - QUIC and HTTP/3

- Added javadocs.
- Fixed race condition in processDataDemand().

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Issue #6728 - QUIC and HTTP/3

- Fixed computation of the last flag.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Issue #6728 - QUIC and HTTP/3

- More javadocs.
- Fixed QuicSession.fill() to also flush() to send to the remote peer the information that the data has been consumed.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet commented:
@gregw updated the comment above.

lorban committed:
Upgrade to quiche 0.10.0

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

sbordet committed:
Issue #6728 - QUIC and HTTP/3

- Avoid sending the last=true flag as a separate QUIC frame.
Apparently Quiche has problem notifying this frame to the receiving peer.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Issue #6728 - QUIC and HTTP/3

- Fixed module-info.java for HTTP/3 modules.
- Updated package-info.java for QUIC.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

lorban committed:
Fix application protocol to make it work with utf8

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

sbordet committed:
Issue #6728 - QUIC and HTTP/3

- Initial support for closing HTTP/3 sessions and session failure events.
- Enforced HTTP/3 frame sequence.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Issue #6728 - QUIC and HTTP/3

- Fixed charset encoding handling.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

lorban committed:
cleanup jna mappings

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

sbordet committed:
Issue #6728 - QUIC and HTTP/3

- Added test case for blocking I/O reads.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Issue #6728 - QUIC and HTTP/3

- Implemented support for GREASE for unidirectional and bidirectional streams.
- Implemented UnknownBodyParser.
- Strengthened validation of frames in control and message streams.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Issue #6728 - QUIC and HTTP/3

- Fixed computation of the last flag.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Issue #6728 - QUIC and HTTP/3

- Fixed module-info.java canonical ordering.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Issue #6728 - QUIC and HTTP/3

- Implemented stream idle timeouts.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Issue #6728 - QUIC and HTTP/3

- Avoid leaking stream instances.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Issue #6728 - QUIC and HTTP/3

- Added more configuration properties.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Issue #6728 - QUIC and HTTP/3

- Implemented GOAWAY parsing/generation.
- Implemented handling of GOAWAY frames.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Issue #6728 - QUIC and HTTP/3

- Implemented goAway mechanism.
- Implemented idle timeout mechanism.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Issue #6728 - QUIC and HTTP/3

- Fixed configuration of QuicConnection.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

lorban committed:
prototype HTTP3 upper layers

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
implement adaptive execution strategy for simple requests

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

sbordet committed:
Issue #6728 - QUIC and HTTP/3

- Implemented component hierarchy.
- Implemented dumpability of components.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Issue #6728 - QUIC and HTTP/3

- Implemented graceful shutdown functionality.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Issue #6728 - QUIC and HTTP/3

- Clearly separated QUIC errors from HTTP3 errors.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Issue #6728 - QUIC and HTTP/3

- Updated to 10.0.8-SNAPSHOT.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Issue #6728 - QUIC and HTTP/3

- Implemented better toString() for Instruction implementations.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Issue #6728 - QUIC and HTTP/3

- Optimized generation of HEADERS frames.
- Changed QPACK encoder to not flip the buffer, necessary for the optimization above.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Issue #6728 - QUIC and HTTP/3

- Improved configuration of client and server.
- Implemented handling of QPACK exceptions.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

lorban committed:
WIP: implementation of server's upper layers input and output

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
WIP: reworking H3 channel impl + workaround quiche bug by using CUBIC congestion control

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

sbordet committed:
Issue #6728 - QUIC and HTTP/3

- Improved configuration of client and server.
- Started implementation of HttpClientTransportOverHTTP3.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Issue #6728 - QUIC and HTTP/3

- WIP on the threading model.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Issue #6728 - QUIC and HTTP/3

- Improved javadocs and formatting.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Issue #6728 - QUIC and HTTP/3

- WIP on the client upper layer.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Issue #6728 - QUIC and HTTP/3

- Improvements to the thread model implementation.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Issue #6728 - QUIC and HTTP/3

- WIP on the client upper layer.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Issue #6728 - QUIC and HTTP/3

- Removed dataFrame queue in HTTP3StreamConnection.
- Made readData() idempotent.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Issue #6728 - QUIC and HTTP/3

- Moved ExecutionStrategy from ServerQuicConnection to QuicConnection.
  For the server the produced task is declared as BLOCKING, but for the
  client the produced task is NON_BLOCKING.
- Fixed race condition in QuicSession.process(...).
- Updated quic-quiche pom.xml.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Issue #6728 - QUIC and HTTP/3

- Added timeouts for QUIC tests.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Issue #6728 - QUIC and HTTP/3

- QPACK exceptions now use long instead of int, to be consistent with other error codes.
- Fixed ManagedSelector to count down the stop latches in finally blocks, so that they are always counted down even in case of exceptions.
- Improved exception handling in case of closes.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Issue #6728 - QUIC and HTTP/3

- Improved logging in QuicheConnection.
- Incremented timeout for echo test in case of large content.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Issue #6728 - QUIC and HTTP/3

- Made sure Connection.Listeners are properly notified.
- Fixed removal of QuicStreamEndPoints from QuicSession in case of successful request/response exchanges.
- Avoid spurious wakeups by shutting down input after receiving a frame with last=true.
- Updated HttpClient transport tests to work with UNIX_DOMAIN.
- Started updating HttpClient transport tests to work with HTTP/3.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

lorban committed:
Issue #6728 - QUIC and HTTP/3
 - Renamed methods & variables for clarity
 - Tightened code visibility

lorban committed:
Issue #6728 - QUIC and HTTP/3
 - Add server UML seq. diagram from network to Handler

sbordet committed:
Issue #6728 - QUIC and HTTP/3

- More work on making HTTP semantic layer work on top of HTTP/3.
- Various fixes and improvement to HTTP client transport tests.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Issue #6728 - QUIC and HTTP/3

- More fixes and improvement to HTTP client transport tests.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Issue #6728 - QUIC and HTTP/3

- More fixes and improvement to HTTP client transport tests.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Issue #6728 - QUIC and HTTP/3

- More fixes and improvement to HTTP client transport tests.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Issue #6728 - QUIC and HTTP/3

- Made BlockingContentProducer.onContentProducible() idempotent by checking if the input
is unready before releasing the semaphore.
This is necessary because HTTP/3 will call onContentProducible() just after receiving the request,
while other protocols assume that content is producible and only call onContentProducible()
when they read all the available content.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Issue #6728 - QUIC and HTTP/3

- Made HttpChannelOverHTTP3.needContent() to look for content if none is immediately available.
- Improved javadocs.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Issue #6728 - QUIC and HTTP/3

- Fixed 100 Continue handling.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Issue #6728 - QUIC and HTTP/3

- Fixed UNIX_DOMAIN tests.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Issue #6728 - QUIC and HTTP/3

- More fixes and improvement to HTTP client transport tests.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Issue #6728 - QUIC and HTTP/3

- Fixed buffer release to avoid leaks.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Issue #6728 - QUIC and HTTP/3

- Improved close mechanism.
Now error and reason are propagated at the HTTP/3 level, in case e.g. applications want to take statistics about the error codes.
- Improved buffer handling to be sure they are properly released back to the pool.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Issue #6728 - QUIC and HTTP/3

- Fixed end of stream detection.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Issue #6728 - QUIC and HTTP/3

- Fixed HttpChannelOverHTTP3 input reading.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Issue #6728 - QUIC and HTTP/3

- Fixed trailer frame handling.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Issue #6728 - QUIC and HTTP/3

- Fixed idle timeout handling.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Issue #6728 - QUIC and HTTP/3

- Fixed test case to work more consistently with HTTP/3.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Issue #6728 - QUIC and HTTP/3

- Fixed max streams semantic for HTTP/3.
It's not *concurrent* streams but *cumulative* streams that are limited.
Fixed MultiplexConnectionPool to take that into account with the introduction
of ConnectionPool.MaxUsable.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Issue #6728 - QUIC and HTTP/3

- More test fixes.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Issue #6728 - QUIC and HTTP/3

- Fixed parsing of HEADERS frames.
- Fixed locking in QpackEncoder.
- Fixed creation of QuicStreamEndPoints.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Issue #6728 - QUIC and HTTP/3

- Incremented test timeouts and decremented runs to allow H3 to finish.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Issue #6728 - QUIC and HTTP/3

- Fixed session failure notification.
Now the failure is propagated to listeners.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Issue #6728 - QUIC and HTTP/3

- Incremented test timeouts.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

lorban committed:
Issue #6728 - QUIC and HTTP/3
 - add windows and mac binding support

lorban committed:
Issue #6728 - QUIC and HTTP/3
 - use proper constant

lorban committed:
Issue #6728 - QUIC and HTTP/3
 - use proper parent pom

sbordet committed:
Issue #6728 - QUIC and HTTP/3

- Try to pinpoint the elusive failures.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Issue #6728 - QUIC and HTTP/3

- Renamed Panama/JDK to Foreign/Incubator.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Issue #6728 - QUIC and HTTP/3

- Fixed missing <release>17</release> element.
- Fixed --add-opens option in Surefire to run the tests.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Issue #6728 - QUIC and HTTP/3

- Trying to replicate test failures.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Issue #6728 - QUIC and HTTP/3

- Trying to replicate test failures.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Issue #6728 - QUIC and HTTP/3

- Trying to replicate test failures.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Issue #6728 - QUIC and HTTP/3

- Trying to replicate test failures.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Issue #6728 - QUIC and HTTP/3

- Trying to replicate test failures.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Issue #6728 - QUIC and HTTP/3

- Trying to replicate test failures.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

lorban committed:
Issue #6728 - QUIC and HTTP/3
 - foreign: delete native lib copy on exit

lorban committed:
Issue #6728 - QUIC and HTTP/3
 - commonalize quiche constants across bindings

lorban committed:
Issue #6728 - QUIC and HTTP/3
 - refactor resource scopes and memory segments usage

sbordet committed:
Issue #6728 - QUIC and HTTP/3

- Trying to replicate test failures.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Issue #6728 - QUIC and HTTP/3

- Giving up on trying to get Jenkins run UDP/QUIC/HTTP3 load tests.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

lorban committed:
Issue #6728 - QUIC and HTTP/3
 - add mac amd64/arm64 and windows amd64 native lib

lorban committed:
Issue #6728 - QUIC and HTTP/3
 - JNA seems to require the module to be wide open

lorban committed:
Issue #6728 - QUIC and HTTP/3
 - re-order native helper static loader

sbordet committed:
Issue #6728 - QUIC and HTTP/3

- Test logging at INFO level.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Issue #6728 - QUIC and HTTP/3

- Larger timeout for file upload test.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Issue #6728 - QUIC and HTTP/3

- Incremented test timeouts.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

olamy committed:
remove native libraries and use external project

Signed-off-by: Olivier Lamy <oliver.lamy@gmail.com>

lorban committed:
Issue #6728 - QUIC and HTTP/3
 change quiche native dependency to released version

sbordet committed:
Issue #6728 - QUIC and HTTP/3

- Split Stream in Stream.Client and Server.Server, so segregate client-specific actions and events.
Now, only Stream.Server has method respond(), and only Stream.Client.Listener has method onResponse().
- Improved javadocs, and updated javadoc module to create javadocs for both http3 and quic modules.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Issue #6728 - QUIC and HTTP/3

- Improved code comments.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Issue #6728 - QUIC and HTTP/3

- Removed Java 17 options, not necessary anymore because we don't javadoc the quic-quiche module.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Issue #6728 - QUIC and HTTP/3

- Fixed javadoc error.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Issue #6728 - QUIC and HTTP/3

Updates after merge of #7100.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Issue #6728 - QUIC and HTTP/3

Added http3 Jetty module and distribution test.
Implemented simple logic to send the Alt-Svc header in HTTP/2 responses.
Updated JNA dependency to use jna-jpms.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Issue #6728 - QUIC and HTTP/3

Tagged the `http3` Jetty module as "experimental".

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet commented:
@gregw things to review more carefully:

* HttpClientTransportDynamic.findConnector(): Avoid to always allocate a default `ClientConnector` and searching deep in the hierarchy, as the `ClientConnector` may now be part of a lower level component.
* `MultiplexConnectionPool`: now also `getMaxUsageCount()` is overridden to ask the connection, if it implements `MaxUsable` to return the count. This is necessary because in HTTP/3 a connection can only open N streams (differently from HTTP/1 or HTTP/2 where a connection could be reused almost infinitely, HTTP/3 limits the number of requests you can make on a connection, and then the client must "open" another connection -- or rather, allocate a new connection ID).
* `PreEncodedHttpField`: removed the hardcoded number of encoders.
* `AsyncContentProducer`: added `isUnready()`, used in `BlockingContentProducer`.

The rest is either new code or tests modified to run HTTP/3 as well.

sbordet committed:
Issue #6728 - QUIC and HTTP/3

Removed obsolete TODO comment.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Issue #6728 - QUIC and HTTP/3

Made QuicConnection's fillable callback of type EITHER.
This is necessary when there are no threads available,
because we still want to read from the network and feed
Quiche so that it can progress with respect to packet
acknowledgments and flow control updates.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Issue #6728 - QUIC and HTTP/3

Made ServerProtocolSession's producer task of type EITHER.
This is necessary to allow for example DATA frames to be
processed to unblock blocked reads.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

gregw commented on jetty-http3/http3-server/src/main/config/modules/http3.mod:
```diff
@@ -0,0 +1,50 @@
+[description]
+Enables the support for the HTTP/3 protocol.
```
```suggestion
Enables experimental support for the HTTP/3 protocol.
```

gregw commented on jetty-http3/http3-server/src/main/java/org/eclipse/jetty/http3/server/HTTP3ServerConnector.java:
```diff
@@ -0,0 +1,61 @@
+    @Override
+    protected void doStart() throws Exception
+    {
+        super.doStart();
```
```suggestion
        LOG.info("H3/Quic support is experimental");
        super.doStart();
```

gregw commented on jetty-http3/http3-server/src/main/java/org/eclipse/jetty/http3/server/HTTP3ServerConnector.java:
```diff
@@ -0,0 +1,61 @@
+/**
+ * <p>A HTTP/3 specific {@link QuicServerConnector} that configures QUIC parameters according to HTTP/3 requirements.</p>
+ */
+public class HTTP3ServerConnector extends QuicServerConnector
```
I think having this class is not correct long term.   Just like an SslConnector can be h1 or h2, a QuicServerConnector can be h1, h2 or h3.
So don't we want to just have the QuicServerConnector and use code/xml to add the supported protocols to it? 

OK for it to stay as an expediency for the experimental h3, but we should review the need for it.

gregw commented:
So long as this is clearly marked as experimental, then I think it is ok to merge. 
We need to have an architecture review of the approach after we have settled on some more ideas in jetty-12

sbordet committed:
Issue #6728 - QUIC and HTTP/3

Added more logging about HTTP/3 being experimental.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet commented:
@gregw I have incorporated your suggestions and added a log on the client too.

sbordet committed:
Issue #6728 - QUIC and HTTP/3

Fixed javadocs.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

gregw commented:
@sbordet so since this is "my" PR, I think you can review it and we can merge.  The only permanent thing from this is the maven coords and they are reasonable going forward anyway.

sbordet commented:
I'll merge - we have done some improvements on the module system but we're waiting for libraries to be in Maven Central.

sbordet committed:
Issue #6728 - QUIC and HTTP/3

Split the http3 module, introducing the jna and quiche modules with their versions and licenses.
Updated project to reference the new groupId of the quiche binaries.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Issue #6728 - QUIC and HTTP/3

Fixed test by approving licenses in DistributionTests.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Issue #6728 - QUIC and HTTP/3

Documentation.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>
